### PR TITLE
add guide for gang scheduling with RayJob and Kueue

### PR DESCRIPTION
**Pull request by andrewsykim**
## Why are these changes needed?

Per discussion in https://github.com/ray-project/kuberay/issues/1890, we would like to add a guide on how to use RayJob and Kueue for gang scheduling. 

## Related issue number

https://github.com/ray-project/kuberay/issues/1890

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by andrewsykim** [comment]
@kevin85421 this is ready for an initial review

**Comment by andrewsykim** [comment]
@angelinalg can you help review this PR please? It is similar to https://github.com/ray-project/ray/pull/42903

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:207]: However, when I increase the number of nodes in the `default-pool` from 1 to 2, it provisions a new node within 30 seconds.

Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:207]: This seems to be a bug for the beta feature `gcloud beta container node-pools create gpu-node-pool ...`.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (kevin85421) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:243]: Perhaps we should move it to the beginning of the document. "Why" is more important than "How".

**Comment by kevin85421** [comment]
The following process will be similar to https://github.com/ray-project/ray/pull/42903#issuecomment-1925205038.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:84]: Upon further consideration, we don't need to add the YAML to the KubeRay repository if we decide to paste and explain the entire YAML in this document. Keeping the consistency between the YAML and the YAML in the doc is not easy.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:207]: > My provisioning request has already been created for 15 minutes, but PROVISIONED is still False.

This is expected IF there's no GPU capacity. It's possible there was no GPUs at the time you created the RayJob but then some became available when you tried to manually scale the node pool. 

> Is there any way to know the progress of the provisioning process? 

I believe there is some work planned to improve the visibility of ProvisioningRequest but I'll need to follow-up with Kueue folks on it

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:243]: Good point, I'll move it

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:116]: Good catch, we actually don't need preemption in this example

Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:116]: When pre-empting, yes it will set Suspend=true

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:198]: This example was based on v1.0.0

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:212]: > This is expected IF there's no GPU capacity. It's possible there was no GPUs at the time you created the RayJob but then some became available when you tried to manually scale the node pool.

You may have misunderstood my messages slightly. Let me summarize why I think it is a bug.

* `default-pool` is created by `gcloud container clusters create` (not a beta feature).
* `gpu-node-pool` is created by `gcloud beta container node-pools ...` (a beta feature).
* My provisioning request has already been created for 15 minutes, but PROVISIONED is still False.
* I tried to manually scale up the size of `gpu-node-pool` (https://github.com/ray-project/ray/pull/43062#discussion_r1491718392) from 0 to 1, but nothing happened (the GKE Web UI still shows the node pool has 0 node).
* I tried to manually scale up the size of `default-pool` from 1 to 2. The GKE Web UI shows the node pool has 2 nodes immediately before the new node is actually created.
* Hence, I expect that the web UI should change immediately when I edit it manually.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:203]: Maybe we can update the status?

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:212]: > My provisioning request has already been created for 15 minutes, but PROVISIONED is still False.

PROVISIONED=False in the the ProvisioningRequest right? ProvisioningRequest does not pertain to the node pool itself, it pertains the set of nodes that will be used to run the RayJob. Provisioned=True means the nodes in the node pool have been created. 

`gpu-node-pool is created by gcloud beta container node-pools ... (a beta feature).` succeeds quickly because you create a node pool with 0 nodes.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:111]: @kevin85421 yes that's correct, Kueue is always in gang mode in that it will only admit workloads when all of it's resources are available.

However, without autoscaling and provisioning request, it is up to the cluster admin to properly configure quotas in ClusterQueue to reflect the capacity available in the cluster. With autoscaling and provisioning request, gang scheduling is automatically handled and the quota can be configured to "infinity". Both approaches work and have trade-offs, for this guide I decided to use the autoscaling approach so we don't have to configure static quotas

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:78]: It's (1). Kueue can initiate scale up of nodes using ProvisioningRequest. However, the difference from normal scale up is that ProvisioningRequest blocks node scale up until all resources are available at once

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:56]: done

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:216]: > Is there any way to know the progress of the provisioning process?

Regarding this, I spoke with @alculquicondor and he opened this issue as a follow-up https://github.com/kubernetes-sigs/kueue/issues/1749

**Review by kevin85421 — APPROVED** [review]
Stamp. The following process will be the same as https://github.com/ray-project/ray/pull/42903#issuecomment-1925205038.
* Merge this PR after the Ray doc team approves.
* I will open a follow-up PR to improve it if needed.

**Comment by kevin85421** [comment]
I have already requested Ray doc team to review this PR.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:35]:
```suggestion
Create a GKE cluster with `enable-autoscaling` option:
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:42]:
```suggestion
Create a GPU node pool with the `enable-queued-provisioning` option:
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:57]:
```suggestion
This command creates a node pool which initially has zero nodes. Use the `gcloud beta` command because some flags have beta status.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:58]:
```suggestion
The `--enable-queued-provisioning` flag enables "queued provisioning" in the Kubernetes node autoscaler using the ProvisioningRequest API. More details are below.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:59]:
```suggestion
You need to use the `--reservation-affinity=none` flag because GKE doesn't support Node Reservations with ProvisioningRequest.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:62]:
```suggestion
"enable-queued-provisioning" is only available on versions 1.28+ with the `gcloud beta` command
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:66]:
```suggestion
## Install the KubeRay operator
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:83]:
```suggestion
1. Dynamically adding new nodes to the cluster when a job needs more resources.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:84]:
```suggestion
2. Blocking the admission of new jobs that are waiting for sufficient resources to become available.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:88]:
```suggestion
### Create Kueue resources
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:91]:
```suggestion
* [ClusterQueue](https://kueue.sigs.k8s.io/docs/concepts/cluster_queue/): Defines quotas and fair sharing rules
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:92]:
```suggestion
* [LocalQueue](https://kueue.sigs.k8s.io/docs/concepts/local_queue/): A namespaced queue, belonging to a tenant, that references a ClusterQueue
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:93]:
```suggestion
* [ResourceFlavor](https://kueue.sigs.k8s.io/docs/concepts/resource_flavor/): Defines what resources are available in the cluster, typically from Nodes
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:94]:
```suggestion
* [AdmissionCheck](https://kueue.sigs.k8s.io/docs/concepts/admission_check/): A mechanism allowing components to influence the timing of a workload admission
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:128]:
```suggestion
  namespaceSelector: {} # match all
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:158]:
```suggestion
This example configures Kueue to orchestrate the gang scheduling of GPUs. However, you can use other resources such as CPU and memory.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:170]:
```suggestion
to assign the RayJob to the LocalQueue that you created earlier:
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:186]:
```suggestion
Following is the expected behavior when you deploy a GPU-requiring RayJob to a cluster that initially lacks GPUs:
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:191]:
```suggestion
* Kueue admits the RayJob, schedules Ray nodes on the newly provisioned nodes, and the RayJob execution begins.
```

Comment (angelinalg) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:191]: Does Kueue schedule the Ray nodes? If not, I need to correct this rewrite.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:193]:
```suggestion
If GPUs are unavailable, Kueue keeps suspending the RayJob. In addition, the node autoscaler avoids provisioning new nodes until it can fully satisfy the RayJob's GPU requirements.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:195]:
```suggestion
Upon creating a RayJob, notice that the RayJob status is immediately `suspended` despite the ClusterQueue having GPU quotas available.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:213]:
```suggestion
Kueue keeps suspending this RayJob until its corresponding ProvisioningRequest is satisfied.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:214]:
```suggestion
List ProvisioningRequest resources and their status with this command:
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:221]:
```suggestion
Note the two coloumns in the output: `ACCEPTED` and `PROVISIONED`.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:222]:
```suggestion
`ACCEPTED=True` means that Kueue and the Kubernetes node autoscaler have acknowledged the request.
```

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:224]:
```suggestion
Once both of these conditions are true, the ProvisioningRequest is satisfied.
```

**Review by angelinalg — COMMENTED** [review]
Just some nits.

File: doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/examples/rayjob-kueue-gang-scheduling.md:191]: Reworded to:
```
Kueue admits the RayJob, allowing Kubernetes to schedule the Ray nodes on the newly provisioned nodes, and the RayJob execution begins.
```

**Comment by andrewsykim** [comment]
Thanks for the review @angelinalg, addressed your comments!

**Review by angelinalg — APPROVED** [review]

**Pull request merged**
